document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -263,8 +269,166 @@ public BulkUpdateResult bulkUpdateSubDocs(Map<Key, Map<String, Document>> docume
   }
 
   @Override
-  public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest request) {
-    throw new UnsupportedOperationException();
+  public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest request)
+      throws Exception {
+    Operation operation = request.getOperation();
+    switch (operation) {
+        // todo: All of these implementations do a GET -> in memory update -> UPSERT which is not
+        // optimal. Changes this logic to use a single query using jsob_set and json_insert
+      case ADD:
+        return addImpl(request);
+      case SET:
+        return setImpl(request);
+      case REMOVE:
+        return removeImpl(request);
+      default:
+        throw new UnsupportedOperationException("Unsupported operation: " + operation);
+    }
+  }
+
+  private BulkUpdateResult removeImpl(BulkArrayValueUpdateRequest request) throws IOException {
+    Set<Key> keys = request.getKeys();
+    String[] pathTokens = request.getSubDocPath().split("\\.");
+    Map<Key, Document> upsertMap = new HashMap<>();
+    for (Key key : keys) {
+      Query getQuery = getSelectDocumentQueryForKey(key);
+      CloseableIterator<Document> searchRes = search(getQuery);
+      if (searchRes.hasNext()) {
+        JsonNode rootNode;
+        try {
+          Document doc = searchRes.next();
+          rootNode = MAPPER.readTree(doc.toJson());
+        } catch (JsonProcessingException e) {
+          LOGGER.error("Malformed JSON for key: {}", key);
+          continue;
+        }
+        JsonNode currNode = rootNode;
+        for (String pathToken : pathTokens) {
+          currNode = currNode.path(pathToken);
+        }
+        // if we do not end up in an array node, then the path doesn't exist. So do not do anything
+        if (!currNode.isArray()) {
```

Comment:
Can `currNode` be null if the path doesn't exist?

Reply:
`.path` never returns null, it returns a `MissingNode` instead. So this condition will evaluate to be false if path doesn't exist.